File: 03/challenge-03.hs
import Data.List.Split


splitPassword :: String -> (Int, Int, Char, String)
splitPassword "" = (0, 0, ' ', "")
splitPassword s = do
    let parts = splitOn " " s
    let range = splitOn "-" (head parts)
    let min = read (head range) :: Int
    let max = read (range !! 1) :: Int
    let char = head (parts !! 1)
    let password = parts !! 2
    (min, max, char, password)

checkValidity :: (Int, Int, Char, String) -> Bool
checkValidity (min, max, char, password) = do
    let count = length $ filter (==char) password
    count >= min && count <= max

fourth :: (a, b, c, d) -> d
fourth (_, _, _, d) = d

main :: IO()
main = do
  let filename = "./03/in"
  contents <- readFile filename
  let lines = splitOn "\n" contents
  let passwords = map splitPassword lines
  let invalidPasswords = filter (not . checkValidity) passwords
  print "Invalid password #42:"
  print (fourth (invalidPasswords !! 41))

File: 03/challenge-03.ts
import { readFileSync } from 'node:fs'

function isValidKey (key: string, min: number, max: number, char: string): boolean {
  const count = key.split('').filter(ch => ch === char).length
  return count >= min && count <= max
}

function main (): void {
  console.log('Challenge #3')
  const file = readFileSync('./03/in').toString()
  const lines = file.split('\n')
  lines.pop()
  const validKeys = lines.filter((line, index) => {
    const [policy, key] = line.split(':')
    const [minMax, char] = policy.split(' ')
    const [min, max] = minMax.split('-').map(Number)
    const isValid = isValidKey(key, min, max, char)
    return !isValid
  })
  console.log(`Unvalid key #42: ${validKeys[41]}`)
}

main()
